crates/modules/network-monitor/probes.bpf.c

```diff
@@ -312,6 +313,27 @@ static __always_inline void on_accept_exit(void *ctx, long ret) {
   output_network_event(ctx, event);
 }
 
+#define ITER_UBUF 5
+
+static __always_inline const struct iovec
+*iter_iov(const void *msg_iter) {
+  // Definition of `struct iov_iter` used in new kernels (>=6.4).
+  const struct iov_iter *msg_iter_nocompat = msg_iter;
+
+  // Check if `msg_iter` matches the new definition of `struct iov_iter`. The
+  // `__ubuf_iovec` field doesn't exist in kernels <= 6.4.
+  if (bpf_core_field_exists(msg_iter_nocompat->__ubuf_iovec)) {
+    if (BPF_CORE_READ(msg_iter_nocompat, iter_type) == ITER_UBUF) {
+      return &msg_iter_nocompat->__ubuf_iovec;
+    }
+    return BPF_CORE_READ(msg_iter_nocompat, __iov);
```

Comment:
why in the first case the access is plain and only in the other case you use `BPF_CORE_READ`?

Reply:
`__ubuf_iovec` and `__iov` have different types, the former being `struct iovec` and the latter - `struct *iovec`. Doing `&BPF_CORE_READ` is impossible.

But thanks for pointing it out. I probably need to think of some better way of handling these two cases. And return a pointer only when we are supposed to use `__iov`, not `__ubuf_iovec`. I'll think about it.

Reply:
Actually, I found a clean solution - returning `iov_iter` directly in that new function solves the problem entirely (I can just do `BPF_CORE_READ(msg_iter_nocompat, __ubuf_iovec.iov_base)`).

This is how I'm doing it now (I pushed the change):

https://github.com/Exein-io/pulsar/blob/1f53a1916aa93d0d1c312e35f6f2de5d47cc7209/crates/modules/network-monitor/probes.bpf.c#L318-L335

Let me know if you agree with that.

Reply:
yeah, now it looks great.

When I saw the arrow operator on the kernel type, it raised me the alarm. I didn't even notice the different types